Use auth middleware in trigger service `/v1/start` endpoint

aherrmann-da opened the pull request:
- Adds a `state` parameter to the auth middleware's `/login` endpoint. Used by the trigger service to register a continuation on its callback endpoint.
- Adds an optional (and for now hidden) `--auth` parameter to the trigger service. This can be used to point to the auth middleware.
- Implement an `authorize` directive in the trigger service. When `--auth` is defined then this directive will require authentication/authorization on the middleware with the necessary claims. It goes through the auth/login flow as described in `authentication.md`. If a login is required then it stores the current request to resume once login was completed.
- Forwards the token from the auth middleware to the ledger client configuration for the newly started trigger.
- Removes the `applicationId` from the token issued by the example auth0 rule. Needed because the trigger service assigns a new application id to every running trigger.
- Similarly, make `applicationId` optional on the test oauth2 server.
- Add a trigger service test suite that uses that auth service on the `/v1/start` endpoint.
  - Needs to track cookies between http requests and follow redirects for the auth middleware to work.
  - Needs a ledger client with a JWT sufficient to perform various ledger commands during the test suite.

I have tested this against sandbox classic and auth0 with the heartbeat test trigger using the following set of commands.
```
$ bazel run //ledger/sandbox-classic:sandbox-classic-binary -- -a localhost -p 6865 --ledgerid daml-auth0-example-ledger -w --auth-jwt-rs256-jwks https://XYZ.eu.auth0.com/.well-known/jwks.json &
$ DAML_CLIENT_ID=XYZ DAML_CLIENT_SECRET=XYZ bazel run //triggers/service/auth:oauth-middleware-binary -- --port 3000 --oauth-auth https://XYZ.eu.auth0.com/authorize --oauth-token https://XYZ.eu.auth0.com/oauth/token &
$ bazel run //triggers/service:trigger-service-binary -- --ledger-host localhost --ledger-port 6865 --auth http://localhost:3000 &

$ curl -X POST localhost:8088/v1/upload_dar -F "dar=@bazel-bin/triggers/tests/acs.dar" -H "Content-type: multipart/form-data" -H "Accept: application/json"
$ http -v --follow :8088/v1/start Content-type:application/json Accept:application/json triggerName=b4f208ad47c15afd8b24d164fdfe5307e17209c01dc4dcc373d234f6c015e303:Heartbeat:test party=Alice --session trigger
```
The latter requires the auth0 session cookies to be stored in the httpie session `trigger`.

This does not yet persist the token in the trigger service's database.
This does not yet handle access token expiry and refresh.
For now only `/v1/start` requires authorization.

Part of #6923

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

cocreature commented on triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:
```diff
@@ -171,20 +188,102 @@ class Server(
+                AuthRequest.Claims(actAs = List(Ref.Party.assertFromString(params.party.toString)))
+              // TODO[AH] Why do we need to pass ec, system explicitly?
+              authorize(claims)(ec, system) { token =>
+                startTrigger(params.party, params.triggerName, token) match {
```
We obviously also need this on other endpoints but I assume we’re going to add this later.

aherrmann-da replied:
I've started with this one since it is the one where the token is actually required for the interaction with the ledger. AFAIK the others don't require authorization as far as the ledger is concerned. Do we also want to use the auth middleware to double as an authorization service for actions on the trigger service itself? E.g. uploading a DAR or stopping a trigger.

aherrmann-da replied:
Thanks for clarifying. Stop seems straight forward enough, upload sounds like it needs a bit more thinking. I'll leave these for future PRs.

cocreature reviewed (commented):
Looks great! A few comments and I’ll take another look once it’s no longer a draft. Can we write a test for this?

aherrmann-da commented on triggers/service/auth/src/main/scala/com/daml/oauth/middleware/README.md:
```diff
@@ -51,7 +51,6 @@ repository](https://github.com/digital-asset/ex-secure-daml-infra).
       context.accessToken[namespace] = {
         // NOTE change the ledger ID to match your deployment.
         "ledgerId": "2D105384-CE61-4CCC-8E0E-37248BA935A3",
-        "applicationId": context.clientName,
```
:+1: I've opened https://github.com/digital-asset/daml/issues/7671 to track this.

aherrmann-da commented on triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:
```diff
@@ -171,20 +188,102 @@ class Server(
+      requestId: UUID)(implicit ec: ExecutionContext, system: ActorSystem): Route = ctx => {
+    authRequests.remove(requestId) match {
+      case None => complete(StatusCodes.NotFound)(ctx)
+      case Some(req) =>
```
This calls [`remove`](https://www.scala-lang.org/api/2.12.2/scala/collection/concurrent/TrieMap.html#remove(k:K):Option[V]) which removes the item from `authRequests`. Or do you mean there's another memory leak here?

cocreature replied:
duh, not sure how I managed to miss that. Sorry :facepalm:

aherrmann-da reviewed (commented):
> Can we write a test for this?

Yes, I think I can write a test for this using the oauth2 test server.

aherrmann-da commented on triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:
```diff
@@ -171,20 +188,102 @@ class Server(
+          entity(as[StartParams]) {
+            params =>
+              val claims =
+                AuthRequest.Claims(actAs = List(Ref.Party.assertFromString(params.party.toString)))
```
The trigger service is using the `Party` type from `com.daml.ledger.api.refinements.ApiTypes.Party` which doesn't require validation. Aligning the auth middleware to use the same type avoids this issue.

aherrmann-da commented:
I've added a test suite for the trigger service using the auth middleware. The trigger service test suite is already configurable around the database (in-mem or on-disk). I've used the same mechanism to add an axis for no-auth or auth-middleware. I also had to extend the test-suite to track cookies and follow redirections to make this work with the auth middleware.

I've slightly changed the way in which requests are stored in the trigger service for the login workflow's callback. It's now storing a `Route` to complete the original request with the access token rather than just storing the original request. This is to avoid infinite redirection loops with the trigger service. The initial approach just stored the original request and replayed the full route. This means we could attempt a second login flow if the first one didn't yield a sufficient token. If a client just blindly follows redirects, they would get stuck in a loop. I've changed it now to store a `Route` as a continuation. This continuation will not attempt a second login flow, but simply fail if authorization does not succeed after login. If authorization is successful it will continue handling the original request where it stopped.

aherrmann-da commented on triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:
```diff
@@ -171,20 +186,112 @@ class Server(
+  // First asking for a token on the `/auth` endpoint and redirecting to `/login` if none was returned.
+  // If a login is required then this will store the current continuation in `authCallbacks`
+  // to proceed once the login flow completed and authentication succeeded.
+  private def authorize(claims: AuthRequest.Claims)(
```
This directive seems reasonably generic and could potentially be factored out into a client library to be re-used between multiple clients that use the auth middleware.

cocreature reviewed (approved):
Great work, thank you very much!

aherrmann-da commented on triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceConfig.scala:
```diff
@@ -36,6 +39,13 @@ final case class JdbcConfig(
     password: String,
 )
 
+private[trigger] final case class AuthTestConfig(
```
Good catch, I've moved it next to the test fixture where it's used.

aherrmann-da commented on triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala:
```diff
@@ -77,9 +88,43 @@ object TriggerServiceFixture extends StrictLogging {
+          cleanup = () =>
+            for {
+              _ <- oauth.unbind()
+              _ <- middleware.unbind()
```
Fair enough, fixed.
